xdsl/transforms/convert_memref_stream_to_loops.py

```diff
@@ -42,62 +41,109 @@ def _insert_load(
 
             def insert_load(
                 source_index: int,
-                source: SSAValue,
-                affine_map_attr: AffineMapAttr,
                 ind_vars: Sequence[SSAValue],
                 rewriter: PatternRewriter,
                 insertion_point: InsertPoint,
             ) -> SSAValue:
+                source = op.operands[source_index // interleave_factor]
+                affine_map_attr = op.indexing_maps.data[
+                    source_index // interleave_factor
+                ]
                 if source_index >= ins_count:
-                    constant_val = constant_vals[source_index - ins_count]
+                    constant_val = constant_vals[
+                        (source_index - ins_count) // interleave_factor
+                    ]
                     if constant_val is not None:
                         return constant_val
 
                 return _insert_load(
                     source_index,
                     source,
                     affine_map_attr,
-                    ind_vars,
+                    tuple(ind_vars) + extra_dim(source_index),
                     rewriter,
                     insertion_point,
                 )
 
         else:
-            insert_load = _insert_load
+
+            def insert_load(
+                source_index: int,
+                ind_vars: Sequence[SSAValue],
+                rewriter: PatternRewriter,
+                insertion_point: InsertPoint,
+            ) -> SSAValue:
+                source = op.operands[source_index // interleave_factor]
+                affine_map_attr = op.indexing_maps.data[
+                    source_index // interleave_factor
+                ]
+                return _insert_load(
+                    source_index,
+                    source,
+                    affine_map_attr,
+                    tuple(ind_vars) + extra_dim(source_index),
+                    rewriter,
+                    insertion_point,
+                )
+
+        def insert_store(
+            source_index: int,
+            value: SSAValue,
+            ind_vars: Sequence[SSAValue],
+            rewriter: PatternRewriter,
+            insertion_point: InsertPoint,
+        ) -> Operation:
+            nonlocal op
+            index = (source_index + ins_count) // interleave_factor
+            destination = op.operands[index]
+            affine_map = op.indexing_maps.data[index].data
+            if isinstance(destination.type, memref.MemRefType):
+                indices = indices_for_map(
+                    rewriter,
+                    insertion_point,
+                    affine_map,
+                    tuple(ind_vars) + extra_dim(source_index),
+                )
+                store_op = memref.Store.get(value, destination, indices)
+            else:
+                store_op = memref_stream.WriteOp(value, destination)
+            rewriter.insert_op(store_op, insertion_point)
+            return store_op
```

Comment:
Could you put some high level comments here as to what these does, when they are called and specifically the special behaviour regarding the implementation of the interleaving (that in that case it needs to be mapped back to the original memref/stream from the "unrolled" block argument)

Ditto for the load